### DISCUSSION TITLE
Dual graph of a PlanarEmbedding

### DISCUSSION THREAD

**Discussion by JannikIrmai** [networkx/networkx#7747] [Algorithms]
Hi all,

as far as I know, networkx does not implement any algorithm for computing the dual graph of a plane graph. I looked into implementing such an algorithm. My idea was to take as input a [PlanarEmbedding](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.planarity.PlanarEmbedding.html) and output a PlanarEmbedding of its dual graph. However, this is not possible because the PlanarEmbedding is implemented as a DiGraph which does not support multiple edges edges between the same pair of nodes. But in general, the dual graph of a plane graph (even without multi-edges) has multi-edges.

So here are my two ideas to resolve this Problem:
1. Implement a method that computes the dual graph of a PlanarEmbedding but only return the graph as a [MultiGraph](https://networkx.org/documentation/stable/reference/classes/multigraph.html) without the corresponding planar embedding. While this may be sufficient for many purposes the information about the embedding is lost.
2. Change the implementation of the PlanarEmbedding class to support multi graphs. However, I do not know what implications this would have on existing methods that work with planar embeddings. I suppose there is a reason why the PlanarEmbedding class is only implemented for simple graphs. But I do not know this reason.

So here are my questions to the community: 
1. Do you think we should add a method for computing the dual graph of a PlanarEmbedding to networkx?
2. If yes, which of the two ways discussed above is better?
3. Is it prefered to implement such a method as a member of the PlanarEmbedding class or as standalone that take a PlanarEmbedding as input?

Thanks,
Jannik

**Comment by rossbar** [comment]
Here's my two cents:

Yes, an algorithm to compute the dual graph of planar graphs seems like a good fit for networkx, and I don't think this functionality has been implemented yet. However, I would avoid working with the `PlanarEmbedding` object entirely. IMO it's better if a `dual_graph` function accepts normal `networkx.Graph` instances, that way you can do something like:

```python
>>> G = nx.dodecahedral_graph()
>>> G_dual = nx.dual_graph(G)
```

without ever having to explicitly deal with planar embeddings.

**Reply by JannikIrmai** [reply]
The problem with computing the dual graph of a (non embedded) graph is that it is not unique. A planar graph may have multiple planar embeddings and  the dual graphs defined by these embeddings need not be isomorphic (more information in [this](https://en.wikipedia.org/wiki/Dual_graph#Uniqueness) wikipedia article). Therefore, a function that accepts `networkx.Graph` instances is ambiguous and the user had no control over which planar embedding is used. 

This is also why I think it is better if an implementation of such a function returns not only a dual graph but a planar embedding of the dual graph. Because if it were to only return the dual graph, one cannot reconstruct the primal graph from the dual graph in a straight forward way.

**Reply by dschult** [reply]
It probably would be good to allow the user to input a NetworkX graph **or** a PlanarEmbedding. But we can work on the PlanarEmbedding input because adding support for Graph input will just involve creating a planar embedding from it and using that.

**Comment by dschult** [comment]
The history (#3040) suggests that the `PlanarEmbedding` class was created as a tool to help determine planarity, and not much effort was made in that initial attempt to make it generally applicable. (At least beyond noting that a multigraph will be planar when the simple version of the multigraph with single edges is planar.  Actually there are tests to ensure that we can input multigraphs to the planarity checking routines).  But the class itself has since been updated to help drawing and interactive changes in graphs while tracking the PlanarEmbedding. So I suspect approach number 2) is the better approach here.  It might result in an easier to read/maintain version of PlanarEmbedding too (part of what I think @rossbar is referring to when suggesting to avoid the planar embedding class).

I think a `dual_graph` function is a good addition to NetworkX. I haven't jumped into details, but it seems like option 2) is the approach to start with (if you want, start with a subclass of PlanarEmbedding, but it might be easier to just jump in and adapt the main one to work with either). The user-facing interface should be a function. But it would be fine to have that call a method on the Planar Embedding.

I have a question or two for you:
- do we care about the multiedges on the dual graph? What applications would use them?
- do you imagine that people will want to go back and forth often? Or is it typical to find the dual graph and work with that -- not going back and forth?

**Reply by JannikIrmai** [reply]
Thanks for the insights and suggestions! I will try to understand the implementation of the `PlanarEmbedding` class in more detail to judge whether there are any obstacle in supporting multiedges.

Regarding the relevance of multiedges:
- Multiedges are necessary in order to have a one-to-one correspondence between the edges of the primal and dual graph.
One application that comes to mind is the the duality of cuts and cycles: Instead of searching for a minimum cut in the primal graph, one can search for a shortest cycle in the dual graph. See the example at the bottom.
- Regarding the question about going back and forth: I can imagine a scenario where the dual graph (and its planar embedding) is modified and after that one is interested in the resulting primal graph. But I do not have a specific application in mind.

Example:

Depicted below in black is a grid graph (primal graph) and a cut (red dashed line) of that graph. Depicted in blue is the dual graph corresponding to the depicted embedding of the black graph. Notice that the blue graph has multiedges. This is because the faces of the corners of the grid graph share two edges with the outer face. The edges of the cut in the primal graph correspond exactly to the edges of the red cycle in the dual graph. If the multiedges of the dual graph where collapsed into a single edge, the cut of the primal graph would still correspond to a cycle in the dual graph. But then it would no longer be possible to reconstruct the cut based on the cycle because it is not clear which edge of the corner of the primal graph should be cut.

![Screenshot from 2024-11-28 15-59-45](https://github.com/user-attachments/assets/016e56cc-234d-4b2b-8545-ffeb1c563049)

**Comment by boothby** [comment]
I'd propose a somewhat more drastic change: the PlanarEmbedding class should go.  Or, at the least, it needs to be renamed.  At the absolute very least, it is a PlaneEmbedding -- a "planar" graph is a graph which *can be* embedded in the plane, a "plane" graph is a graph which *has been* embedded in the plane and we are clearly dealing with the latter case.  Moreover, there's little terribly special about planarity itself in the class: as far as I can tell, the only such detail is where `check_structure` computes the genus of each component and complains if any isn't zero.

What we have is a nearly-serviceable oriented combinatorial embedding object. Yes, it should derive from MultiDiGraph (if we insist on this representation of combinatorial embedding -- I prefer GEMs, but I digress).  I'd rename `check_structure` to be `components_genus` or similar, thowing errors if the embedding is incomplete, and leaving the opinion that genus=0 out of this particular class.  This suggests a refactor: moving `algorithms/planarity.py` into a new directory (`algorithms/topological/planarity.py`?) and calving off this class to a different file, `algorithms/topological/common.py` or the like.

**Reply by boothby** [reply]
Oh, and, I didn't get to the root question: are we computing a dual graph?  To continue on my persnickety focus on terminology: no, you want to compute a dual embedding.  The underlying graph of an embedding can be had by calling `.to_undirected()`.

**Reply by JannikIrmai** [reply]
I think the term `PlanarEmbedding` is fine as this is used synonymous with "plane graph".

But I like your observation that the current `PlanarEmbedding` class is not specific to planarity, except, that the `check_structure` method checks if the genus is zero. So it makes sense to rename the class to `RotationSystem` or `CombinatorialEmbedding` (as described [here](https://en.wikipedia.org/wiki/Rotation_system)) and move its implementation to another file like you suggest.

And yes, you are right that I am interested in a dual embedding and not just the dual graph. So maybe we can implement a `dual_embedding` (or `dual_rotation_system`) function that takes as input a rotation system and outputs the dual rotation system (I think this is also independent of planarity). Maybe it is even better to implement this as a member function `dual` of the embedding class instead? Then there is no confusion about what the input to the function should be. 

Then we can implement a `dual_graph` function like @dschult suggests that takes a graph as input, computes an embedding, computes the dual of that embedding and then returns the underlying graph of the dual embedding.

Some more detail questions:
- How should the bijection between the edges of the primal and dual structure should be returned or if it should be returned at all? One way would be to add the corresponding edge of the primal to the attribute of the dual edge. 
- Another question is how to deal with disconnected components. In a plane graph disconnected components share the same outer face. However, in a rotation system each connected components has its own orbit as an outer face.
